Frameworks/Foundation/NSScanner.mm

```diff
@@ -575,118 +557,283 @@ - (BOOL)scanUpToString:(NSString*)string intoString:(NSString**)intoString {
+                }
+                break;
+
+            case STATE_MANTISSA:
+                if (unicode == '.') {
+                    if (!scannedPeriod) {
+                        scannedPeriod = true;
+                    } else {
+                        goto done_scanning_hex_double; // Stop scanning once a second period has been scanned
+                    }
+
+                } else if (unicode >= '0' && unicode <= '9') {
+                    if (scannedPeriod) {
+                        mantissa += (mantissaDigitPlace /= 16) * (unicode - '0');
+                    } else {
+                        mantissa = (mantissa * 16) + (unicode - '0');
+                    }
+                    hasValue = true;
+
+                } else if (unicode >= 'a' && unicode <= 'f') {
+                    if (scannedPeriod) {
+                        mantissa += (mantissaDigitPlace /= 16) * (unicode - 'a' + 10);
+                    } else {
+                        mantissa = (mantissa * 16) + (unicode - 'a' + 10);
+                    }
+                    hasValue = true;
+
+                } else if (unicode >= 'A' && unicode <= 'F') {
+                    if (scannedPeriod) {
+                        mantissa += (mantissaDigitPlace /= 16) * (unicode - 'A' + 10);
+                    } else {
+                        mantissa = (mantissa * 16) + (unicode - 'A' + 10);
+                    }
+                    hasValue = true;
+
+                } else if (unicode == 'p' || unicode == 'P') {
+                    state = STATE_P;
+
+                } else {
+                    goto done_scanning_hex_double;
+                }
+                break;
+
+            case STATE_P:
+                if (!hasValue) {
+                    goto done_scanning_hex_double;
+                }
+                if (unicode == '-') {
+                    state = STATE_EXPONENT;
+                    exponentIsNegative = true;
+
+                } else if (unicode == '+') {
+                    state = STATE_EXPONENT;
+
+                } else if (unicode >= '0' && unicode <= '9') {
+                    exponent = (unicode - '0');
+                    state = STATE_EXPONENT;
+
+                } else {
+                    goto done_scanning_hex_double;
+                }
+                break;
+
+            case STATE_EXPONENT:
+                if (unicode >= '0' && unicode <= '9') {
+                    // can stop calculating the change to the exponent once this is false, and let ldexp return 0 or HUGE_VAL
+                    if (exponent <= (exponentIsNegative ? 1023 : 1024)) {
+                        exponent = exponent * 10 + (unicode - '0');
+                    }
+
+                } else {
+                    goto done_scanning_hex_double;
+                }
+                break;
+        }
+    }
+
+done_scanning_hex_double:; // Allows breaking out of the loop from inside the switch
+    if (hasValue) {
+        if (result) {
```

Comment:
you mean besides the exponent getting too large and causing ldexp to return HUGE_VAL or 0?

if the mantissa becomes too precise to represent correctly, the latter (least significant) bits get chopped off during the double operations in STATE_MANTISSA